### ISSUE TITLE
compiler ambiguity handled differently between `switch create` and `install`

### ISSUE THREAD

**Issue by jberdine** [ocaml/opam#3815]
Trying to create a switch for the version of ocaml installed on the host using e.g.:
```
opam switch create snafu 4.07.1
```
fails with
```
[ERROR] Compiler selection '4.07.1' is ambiguous. matching packages: { ocaml-base-compiler.4.07.1,
        ocaml-system.4.07.1 }
```
While the message indicates the problem, it is odd that this ambiguity is automatically resolved  by instead executing:
```
opam switch create snafu --empty && opam install --switch=snafu ocaml.4.07.1
```
Is it intentional that these behave differently?

**Comment by rjbou** [comment]
The opam logic behind those commands is quite different: `opam switch create` does some _magic_ to be able to find the compiler if user specifies the full package compiler, just a version, nothing, etc. ; while `opam install ocaml.4.07.0` will just install `ocaml` package in version `4.07.0`, and then gives that request to the solver. As `ocaml` package depends contains 
```
depends: [
  "ocaml-config"
  "ocaml-base-compiler" {= "4.07.0"} |
  "ocaml-variants" {>= "4.07.0" & < "4.07.1~"} |
  "ocaml-system" {>= "4.07.0" & < "4.07.1~"}
]
```
on of theme is choose by the solver.

**Comment by jberdine** [comment]
Yes, I appreciate that behind the scenes these are quite different. But from an end user point of view, I think that they are the same: they create a switch with a given version of the compiler. The fact that the implementation differences lead to different observable behavior is a source of confusion. Furthermore, that there are two ways to accomplish the same thing is probably suboptimal. This is just a UI feedback point.

**Comment by dra27** [comment]
It's inconsistent with `opam init` too, which prefers `ocaml-system` whenever possible (by a different piece of magic). I think that the config facility in https://github.com/ocaml/opam/pull/2930 might allow the magic to be changed in an OCaml-agnostic way?

**Comment by AltGr** [comment]
The core of the issue is, I believe, that you should actually never do `opam install ocaml` (which means "let an implementation be chosen by the solver at random"), but choose an implementation explicitely instead: `opam install ocaml-base-compiler.xxx`.

I am not sure of how to improve the interface in this regard, and agree this is far from obvious, though...

**Comment by jberdine** [comment]
I should note that in the original post
```
opam switch create snafu 4.07.1
```
is, in actual use, part of a script. This script should be robust and work whether or not the user's system has the requested version of ocaml already installed or not, since this is an aspect that is beyond the scope of the build script.

Stepping back, I think it makes sense to have a way to reliably ask opam to create a new switch with a specified compiler version, building the compiler only if necessary. I suppose that using
```
opam switch create snafu ocaml-base-compiler.4.07.1
```
would be robust, but would rebuild the compiler whether or not it is needed. Using
```
opam switch create snafu ocaml-system.4.07.1
```
would only work in case the user's system happens to have the requested version installed.

I could be overlooking other consequences of the difference between these two commands (?), but perhaps it would work to add an option such as `--rebuild-compiler` so that
```
opam switch create snafu 4.07.1
```
prefers `ocaml-system` if available (analogously to `opam init`) and
```
opam switch create snafu --rebuild-compiler 4.07.1
```
uses `ocaml-base-compiler` while
```
opam switch create snafu --no-rebuild-compiler 4.07.1
```
uses `ocaml-system`, and fails if 4.07.1 is not installed on the host.

But this suggestion is perhaps too ocaml and compiler specific.

**Comment by dra27** [comment]
As it happens, @AltGr and I were discussing a proposed change for opam 2.1/2.2 which although coming from me with Windows in mind, would mean that `opam switch create foo ocaml.4.07.1` would give you a system switch if the system compiler is 4.07.1 and otherwise would build it.
